### DISCUSSION TITLE
Collaborative editing based on yjs

### DISCUSSION THREAD

**Discussion by hughfenghen** [codex-team/editor.js#1892] [Q&A]
Hi, all. 

*Sorry, my English is very bad, so I use bilingual to avoid misunderstanding. Thanks google translate.*
*抱歉，我的英语非常糟糕，所以使用双语，避免误解。感谢谷歌翻译。*

I used [yjs](https://github.com/yjs/yjs) + editor.js to implement a collaborative editing **[demo](https://github.com/hughfenghen/y-editorjs)**, Below is the preview.
我使用[yjs](https://github.com/yjs/yjs) + editor.js实现了一个协同编辑的demo，下面是预览图。

![Kapture 2020-05-16 at 16 15 12](https://user-images.githubusercontent.com/3307051/82137224-bb8f0100-9848-11ea-922d-983ebf866622.gif)

It is currently immature, has many bugs, and does not implement cursor synchronization.
目前它还不成熟，有很多bug，没有实现光标同步。

As you can see from the preview, the updated block will flash (Because the new block is used to replace the old block). Hope that editor.js will have a new API to improve the experience.
从预览图中看到，被更新的block会出现闪烁（因为是使用新的block替换旧的block），希望editor.js会有新的API来改善体验。

https://github.com/codex-team/editor.js/issues/1074#issuecomment-602103628
> However, at the moment we can't provide a method to update a certain block, as data passed to the Tool constructor. So to update it, we need to construct the new Block with updated data and replace the old one. I don't think this approach is the best one, so we need to design a block update flow and describe it with APIs and docs.

In addition, [YUndoManager](https://github.com/yjs/yjs#YUndoManager) may be able to optimize the undo feature. (https://github.com/codex-team/editor.js/issues/518)
另外[YUndoManager](https://github.com/yjs/yjs#YUndoManager)也许可以优化撤销功能。

Related issue: https://github.com/codex-team/editor.js/issues/908, https://github.com/codex-team/editor.js/issues/1040

**Comment by sosie-js** [comment]
Just a thought. I am wondering if it is not possible to avoid to run a node server after discovering the mitm+worker technique
used in [Streamsaver](https://github.com/jimmywarting/StreamSaver.js/blob/master/StreamSaver.js)  to create a relay postMessage between clients avoiding to run a node server but we have to check origin to prevent [attacks](http://www.cs.cornell.edu/~shmat/shmat_ndss13postman.pdf). Following this idea, I fall on the promising [easyXDM](https://github.com/oyvindkinsey/easyXDM#readme)

**Comment by blahah** [comment]
Seems like this could just work over websockets between clients with no need for a server to coordinate the collaboration. Is that what you're saying @sosie-js ?

**Comment by talyguryn** [comment]
thank you. we are working on collaborative editing. will be available in the next major release

**Reply by abasse** [reply]
Hi, can you share a timeline for the next major release?

**Reply by wmonecke** [reply]
Yes, a timeline would be insightful!

**Reply by krisnik** [reply]
@talyguryn are there any updates w.r.t collaborative editing in EditorJS? Thanks in advance.
